src/ERC6538Registry.sol

```diff
@@ -24,22 +68,179 @@ contract ERC6538Registry is IERC6538Registry {
+  /// with `\x19\x01` and hashing the result. It corresponds to the hash signed by the
+  /// https://eips.ethereum.org/EIPS/eip-712[`eth_signTypedData`] JSON-RPC method as part of
+  /// EIP-712.
+  /// @dev The following code is from OpenZeppelin's `MessageHashUtils.sol` file. Permalink:
+  /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e70a0118ef10773457f670671baefad2c5ea610d/contracts/utils/cryptography/MessageHashUtils.sol
+  /// @param domainSeparator The domain separator returned by `_domainSeparatorV4`.
+  /// @param structHash The hash of the struct containing the message data, as defined in
+  /// https://eips.ethereum.org/EIPS/eip-712#definition-of-hashstruct
+  function toTypedDataHash(bytes32 domainSeparator, bytes32 structHash)
+    internal
+    pure
+    returns (bytes32 digest)
+  {
+    /// @solidity memory-safe-assembly
+    assembly {
+      let ptr := mload(0x40)
+      mstore(ptr, hex"1901")
+      mstore(add(ptr, 0x02), domainSeparator)
+      mstore(add(ptr, 0x22), structHash)
+      digest := keccak256(ptr, 0x42)
+    }
+  }
+
+  /// @notice Checks if a signature is valid for a given signer and data hash. If the signer is a
+  /// smart contract, the signature is validated against that smart contract using ERC1271,
+  /// otherwise it's validated using `tryRecover`.
+  /// @dev The following code is from OpenZeppelin's `SignatureChecker.sol` file. Permalink:
+  /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e70a0118ef10773457f670671baefad2c5ea610d/contracts/utils/cryptography/SignatureChecker.sol
+  /// @param signer The address that should have signed the message data.
+  /// @param hash The digest of message data.
+  /// @param signature The signature provided by the registrant.
+  function isValidSignatureNow(address signer, bytes32 hash, bytes memory signature)
+    internal
+    view
+    returns (bool)
+  {
+    (address recovered, RecoverError error,) = tryRecover(hash, signature);
+    return (error == RecoverError.NoError && recovered == signer)
+      || isValidERC1271SignatureNow(signer, hash, signature);
+  }
+
+  /// @notice Checks if a signature is valid for a given signer and data hash. The signature is
+  /// validated against the signer smart contract using ERC1271.
+  /// @dev The following code is from OpenZeppelin's `SignatureChecker.sol` file. Permalink:
+  /// https://github.com/OpenZeppelin/openzeppelin-contracts/blob/e70a0118ef10773457f670671baefad2c5ea610d/contracts/utils/cryptography/SignatureChecker.sol
+  /// @param signer The address that should have signed the message data.
+  /// @param hash The digest of message data.
+  /// @param signature The signature provided by the registrant.
+  function isValidERC1271SignatureNow(address signer, bytes32 hash, bytes memory signature)
+    internal
+    view
+    returns (bool)
+  {
+    (bool success, bytes memory result) =
+      signer.staticcall(abi.encodeCall(IERC1271.isValidSignature, (hash, signature)));
```

Comment:
Why does OZ do the call this way?

Reply:
Is the following way better?
```
  function isValidERC1271SignatureNow(address signer, bytes32 hash, bytes memory signature)
    internal
    view
    returns (bool)
  {
    (bool success, bytes memory result) = signer.staticcall(
      abi.encodeWithSelector(bytes4(keccak256("isValidSignature(bytes32,bytes)")), hash, signature)
    );
    return (
      success && result.length >= 32
        && abi.decode(result, (bytes32))
          == bytes32(bytes4(keccak256("isValidSignature(bytes32,bytes)")))
    );
  }
  ``` 
  Updated in [b2af4c6](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/b2af4c605138ec6f0a8e6361bfa0de28cb140139)

Reply:
This change is worse IMO because `encodeCall` is type-safe, but `encodeWithSelector` is not, therefore, you should always prefer `encodeCall`. Same for preferring `.selector` vs. computing it yourself 😅

I think @apbendi is asking why they do a low-level call instead of a high level call, which is a good question. For reference, here's the Safe implementation of 1271 checks: https://github.com/safe-global/safe-contracts/blob/fc87888576b89167fc48046a8e2a88574fb14c4f/contracts/Safe.sol#L282

Reply:
Gotcha, followed Safe implementation of 1271 and now calling this way:
`(
            IERC1271(registrant).isValidSignature(dataHash, signature)
              == IERC1271.isValidSignature.selector
          )`
          in [a5898bd](https://github.com/ScopeLift/stealth-address-erc-contracts/pull/5/commits/a5898bd3ed0e30960e0f17e686d19a073b0ddb62)